File: phone/acg-ios/ACG/Data/GhTagClient.swift
import Foundation

/// List / read / write `label:tag` issues on a GitHub repo. Anonymous
/// reads; writes need a PAT. Same contract as the Kotlin GhTagClient.
final class GhTagClient {
    let repo: String
    let tokenProvider: () -> String?

    init(repo: String = "teslasolar/aicraftspeopleguild.github.io",
         tokenProvider: @escaping () -> String? = { nil }) {
        self.repo = repo
        self.tokenProvider = tokenProvider
    }

    struct TagIssue: Decodable {
        let number: Int
        let title: String
        let body: String?
        let state: String
        let updated_at: String?
        let comments: Int
    }

    struct TagValue: Codable {
        var value: String?
        var quality: String?
        var type: String?
        var description: String?
        var updated_at: String?
    }

    func listTags() async throws -> [TagIssue] {
        var out: [TagIssue] = []
        var page = 1
        while page < 10 {
            let batch: [TagIssue] = try await get("/repos/\(repo)/issues?labels=tag&state=open&per_page=100&page=\(page)")
            out += batch.filter { $0.title.hasPrefix("tag:") }
            if batch.count < 100 { break }
            page += 1
        }
        return out
    }

    func readTag(_ path: String) async throws -> TagValue? {
        let issues = try await listTags()
        guard let issue = issues.first(where: { $0.title == "tag:\(path)" }) else { return nil }
        return parseBody(issue.body)
    }

    func writeTag(_ path: String, value: String, type: String = "String",
                  quality: String = "good", description: String = "written via ACG iOS") async throws -> String {
        guard let tok = tokenProvider() else {
            throw NSError(domain: "ACG.GhTag", code: 401,
                          userInfo: [NSLocalizedDescriptionKey: "no GitHub token configured"])
        }
        let iso = ISO8601DateFormatter().string(from: Date())
        let tv = TagValue(value: value, quality: quality, type: type, description: description, updated_at: iso)
        let bodyJson = try String(data: JSONEncoder().encode(tv), encoding: .utf8) ?? ""

        let existing = try await allIssues(label: "tag").first { $0.title == "tag:\(path)" }
        if let ex = existing {
            _ = try await request("PATCH", "/repos/\(repo)/issues/\(ex.number)",
                                  body: ["body": bodyJson], token: tok)
            _ = try await request("POST",  "/repos/\(repo)/issues/\(ex.number)/comments",
                                  body: ["body": bodyJson], token: tok)
            return "updated"
        } else {
            _ = try await request("POST", "/repos/\(repo)/issues",
                                  body: ["title": "tag:\(path)", "body": bodyJson, "labels": ["tag"]],
                                  token: tok)
            return "created"
        }
    }

    // MARK: - helpers
    private func allIssues(label: String) async throws -> [TagIssue] {
        var out: [TagIssue] = []
        for state in ["open", "closed"] {
            var page = 1
            while page < 10 {
                let batch: [TagIssue] = (try? await get("/repos/\(repo)/issues?labels=\(label)&state=\(state)&per_page=100&page=\(page)")) ?? []
                out += batch.filter { $0.title.hasPrefix("tag:") }
                if batch.count < 100 { break }
                page += 1
            }
        }
        return out
    }

    private func parseBody(_ body: String?) -> TagValue? {
        guard var t = body?.trimmingCharacters(in: .whitespacesAndNewlines), !t.isEmpty else { return nil }
        if t.hasPrefix("```") {
            if let idx = t.firstIndex(of: "\n") { t = String(t[t.index(after: idx)...]) }
            if t.hasSuffix("```") { t.removeLast(3) }
            if t.hasPrefix("json") { t.removeFirst(4); t = t.trimmingCharacters(in: .whitespaces) }
        }
        return try? JSONDecoder().decode(TagValue.self, from: Data(t.utf8))
    }

    private func get<T: Decodable>(_ path: String) async throws -> T {
        let data = try await request("GET", path, body: nil, token: nil)
        return try JSONDecoder().decode(T.self, from: data)
    }

    @discardableResult
    private func request(_ method: String, _ path: String,
                         body: [String: Any]?, token: String?) async throws -> Data {
        guard let url = URL(string: "https://api.github.com" + path) else { throw URLError(.badURL) }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        req.setValue("acg-ios/0.1", forHTTPHeaderField: "User-Agent")
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, resp) = try await URLSession.shared.data(for: req)
        if let http = resp as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw NSError(domain: "ACG.GhTag", code: http.statusCode,
                          userInfo: [NSLocalizedDescriptionKey: "HTTP \(http.statusCode) \(path)"])
        }
        return data
    }
}

File: phone/acg-android/app/src/main/java/com/aicraftspeopleguild/acg/data/GhTagClient.kt
package com.aicraftspeopleguild.acg.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.jsonArray
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Mirror of bin/acg-mqtt.py / guild/Enterprise/L2/lib/gh_tag.py — the
 * GitHub Issues tag database, but accessed from the phone. Anonymous
 * reads (60 req/hr/IP) + PAT-authed writes. No search API use so we
 * dodge indexing lag — walk label:tag issues by title instead.
 */
class GhTagClient(
    private val repo: String = "teslasolar/aicraftspeopleguild.github.io",
    private val tokenProvider: () -> String? = { null },
) {
    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()
    private val json = Json { ignoreUnknownKeys = true }

    /** List every open tag:* issue by title. */
    @Throws(IOException::class)
    fun listTags(): List<GhIssue> {
        val out = mutableListOf<GhIssue>()
        var page = 1
        while (page < 10) {
            val req = req("GET", "/repos/$repo/issues?labels=tag&state=open&per_page=100&page=$page")
            val body = http.newCall(req).execute().use { r ->
                if (!r.isSuccessful) throw IOException("HTTP ${r.code}")
                r.body?.string().orEmpty()
            }
            val arr: JsonArray = try { json.parseToJsonElement(body).jsonArray } catch (_: Exception) { return out }
            val batch = arr.map { json.decodeFromJsonElement(GhIssue.serializer(), it) }
            out += batch.filter { it.title.startsWith("tag:") }
            if (batch.size < 100) return out
            page++
        }
        return out
    }

    @Throws(IOException::class)
    fun readTag(path: String): TagValue? {
        val issue = listTags().firstOrNull { it.title == "tag:$path" } ?: return null
        return parseBody(issue.body)
    }

    /** Upsert. Finds existing tag:<path> issue (open OR closed) and PATCHes body;
     *  otherwise POSTs a new one. Appends a comment with the new value for history. */
    @Throws(IOException::class)
    fun writeTag(path: String, value: String, type: String = "String",
                 quality: String = "good", description: String = "written via ACG Android"): String {
        val tok = tokenProvider() ?: throw IOException("no GitHub token configured")
        val now = java.time.OffsetDateTime.now(java.time.ZoneOffset.UTC)
            .format(java.time.format.DateTimeFormatter.ISO_INSTANT)
        val bodyJson = buildJsonObject {
            put("value", value); put("quality", quality); put("type", type)
            put("description", description); put("updated_at", now)
        }.toString()

        val existing = listTagsAllStates().firstOrNull { it.title == "tag:$path" }
        return if (existing == null) {
            val payload = buildJsonObject {
                put("title", "tag:$path")
                put("body", bodyJson)
                put("labels", buildJsonArray { add("tag") })
            }.toString()
            val r = http.newCall(req("POST", "/repos/$repo/issues", payload, tok)).execute()
            r.use { if (!it.isSuccessful) throw IOException("POST ${it.code}") }
            "created"
        } else {
            val patch = buildJsonObject { put("body", bodyJson) }.toString()
            http.newCall(req("PATCH", "/repos/$repo/issues/${existing.number}", patch, tok)).execute()
                .use { if (!it.isSuccessful) throw IOException("PATCH ${it.code}") }
            val commentPayload = buildJsonObject { put("body", bodyJson) }.toString()
            http.newCall(req("POST", "/repos/$repo/issues/${existing.number}/comments", commentPayload, tok)).execute()
                .use { /* best-effort comment, ignore body */ }
            "updated"
        }
    }

    private fun listTagsAllStates(): List<GhIssue> {
        val out = mutableListOf<GhIssue>()
        for (state in listOf("open", "closed")) {
            var page = 1
            var done = false
            while (!done && page < 10) {
                val req = req("GET", "/repos/$repo/issues?labels=tag&state=$state&per_page=100&page=$page")
                val body: String? = http.newCall(req).execute().use { r ->
                    if (r.isSuccessful) r.body?.string().orEmpty() else null
                }
                if (body == null) { done = true; continue }
                val arr = try { json.parseToJsonElement(body).jsonArray }
                          catch (_: Exception) { done = true; continue }
                val batch = arr.map { json.decodeFromJsonElement(GhIssue.serializer(), it) }
                out += batch.filter { it.title.startsWith("tag:") }
                if (batch.size < 100) done = true
                page++
            }
        }
        return out
    }

    private fun req(method: String, path: String, body: String? = null, token: String? = null): Request {
        val b = Request.Builder().url("https://api.github.com$path")
            .header("Accept", "application/vnd.github+json")
            .header("User-Agent", "acg-android/0.1")
        token?.let { b.header("Authorization", "Bearer $it") }
        val rb = body?.toRequestBody("application/json".toMediaType())
        return when (method) {
            "GET"   -> b.get().build()
            "POST"  -> b.post(rb ?: "".toRequestBody("application/json".toMediaType())).build()
            "PATCH" -> b.patch(rb ?: "".toRequestBody("application/json".toMediaType())).build()
            else    -> b.method(method, rb).build()
        }
    }

    private fun parseBody(body: String?): TagValue? {
        if (body.isNullOrBlank()) return null
        var t = body.trim()
        if (t.startsWith("```")) {
            t = t.substringAfter("\n", t)
            if (t.endsWith("```")) t = t.substring(0, t.length - 3)
            if (t.trimStart().startsWith("json")) t = t.trimStart().removePrefix("json").trimStart()
        }
        return try { json.decodeFromString(TagValue.serializer(), t) } catch (_: Exception) { null }
    }
}
